apps/token/src/routes/rewards/epoch-individual-rewards/generate-epoch-individual-rewards-list.ts

```diff
@@ -0,0 +1,88 @@
+import { BigNumber } from '../../../lib/bignumber';
+import { RowAccountTypes } from '../shared-rewards-table-assets/shared-rewards-table-assets';
+import type { RewardFieldsFragment } from '../home/__generated__/Rewards';
+import type { AccountType } from '@vegaprotocol/types';
+
+export interface EpochIndividualReward {
+  epoch: string;
+  rewards: {
+    asset: string;
+    totalAmount: string;
+    rewardTypes: {
+      [key in AccountType]?: {
+        amount: string;
+        percentageOfTotal: string;
+      };
+    };
+  }[];
+}
+
+const emptyRowAccountTypes = Object.keys(RowAccountTypes).map((type) => [
+  type,
+  {
+    amount: '0',
+    percentageOfTotal: '0',
+  },
+]);
+
+export const generateEpochIndividualRewardsList = (
+  rewards: RewardFieldsFragment[]
+) => {
+  // We take the rewards and aggregate them by epoch and asset.
+  const epochsMap = rewards.reduce((map, reward) => {
+    const epochId = reward.epoch.id;
+    const assetName = reward.asset.symbol;
+    const rewardType = reward.rewardType;
+    const amount = reward.amount;
+    const percentageOfTotal = reward.percentageOfTotal;
+
+    // if the rewardType is not of a type we display in the table, we skip it.
+    // @todo check if this is the correct behaviour
+    if (!Object.keys(RowAccountTypes).includes(rewardType)) {
```

Comment:
suggestion (non-blocking): if you'd take the keys call out from the reduce, and define it above it as like `const accountTypeKeys = Object.keys(RowAccountTypes)`, it wouldn't need to be called on every iteration. hashtag performance nazi. not a big deal tho, feel free to dismiss

Reply:
Love that, thanks!